scratch/compiler_driver.py

```diff
@@ -0,0 +1,134 @@
+import ast
+from dataclasses import dataclass
+from typing import Dict, Optional, Set, List
+import sys
+
+
+@dataclass
+class SymbolInfo:
+    """Placeholder for symbol information. Will be expanded later."""
+    pass
+
+
+class SymbolTableBuilder(ast.NodeVisitor):
+    """AST visitor that builds a symbol table while traversing the parse tree."""
+
+    def __init__(self):
+        self.symbols = {}  # Global symbol table
+        self.scopes = [{}]  # Stack of scopes, with global scope at index 0
+        self.current_scope = self.scopes[0]
+
+    def visit_Name(self, node):
+        """Process name nodes to track variable references and definitions."""
+        if isinstance(node.ctx, ast.Store):
+            # This is a definition or assignment
+            self.current_scope[node.id] = SymbolInfo()
+            # Also add to global symbol table with fully qualified name
+            self.symbols[node.id] = SymbolInfo()
+        self.generic_visit(node)
+
+    def visit_FunctionDef(self, node):
+        """Process function definitions."""
+        # Register the function name in the current scope
+        self.current_scope[node.name] = SymbolInfo()
+        self.symbols[node.name] = SymbolInfo()
```

Comment:
I'd expect this to have qualified name

Reply:
372d5742a46bc60e4b648185770c02172222a6da